go/kbfs/libkbfs/favorites.go

```diff
@@ -618,23 +652,65 @@ func (f *Favorites) Delete(ctx context.Context, fav favorites.Folder) error {
 	})
 }
 
-// RefreshCache refreshes the cached list of favorites.
-func (f *Favorites) RefreshCache(ctx context.Context) {
+type FavoritesRefreshMode int
+
+const (
+	FavoritesRefreshModeInMainFavoritesLoop = iota
+	FavoritesRefreshModeBlocking
+)
+
+// RefreshCache refreshes the cached list of favorites. If async is true, then
+// this request is processed outside of the normal Favorites queue order.
+//
+// In FavoritesRefreshModeBlocking, request the favorites in this thread,
+// then send them off to the main thread to be processed.
+// In FavoritesRefreshModeInMainFavoritesLoop, this just sets up a request and
+// sends it to the main thread to process it - this is useful if e.g.
+// the favorites cache has not been initialized at all and cannot serve any
```

Comment:
Sorry I still don't understand why it needs to be initialized in this goroutine -- why wouldn't doing it in the favorites loop be good enough?  Seems like in either case it won't be able to serve favorites request until the background loop gets the request and populates the cache.

Does this have anything to do with the timeout, or are they orthogonal issues that you just happen to be addressing in the same PR?

Reply:
The reason to avoid doing it in the main loop is that we want to be able to serve requests from the cache while waiting for the refresh to come from the network.

Reply:
Ohh I see.  It would be great if you could explain that in the comments.

But in that case, there seems to be a problem. You could have multiple sequential `FavoritesChanged` RPCs come in, corresponding to different changes, and if the first one takes longer to complete than the second one, it would overwrite the second one's results in the cache and we'd lose the latest favorites changes.

Seems to me you still have to have some serialization mechanism to make sure there's only one outstanding fetch at a time, and that new refresh requests trigger new API server requests in serial.  For example, there could be a new goroutine managed by `Favorites` that processes requests one at a time and then sends the results over to the main goroutine when each request is complete.

That's assuming we want the property where requests don't block after one of these notifications come in; it's not clear to me that we do.  If we've received a refresh notification, seems like we shouldn't knowingly serve stale data.  A delay in that case seems justified, even in the case where we go offline immediately and it blocks for the full timeout.

Reply:
> Seems to me you still have to have some serialization mechanism to make sure there's only one outstanding fetch at a time

That's what the `refreshWaiting` channel is for - it only lets one refresh happen at a time. There *is* a possible race here:

(1) Alice does something to invalidate Bob's cache
(2) Bob initiates a refresh
(3a) Bob's request is processed by the server. 
(4a) Carly does something to invalidate Bob's cache
(4b) Bob gets Carly's invalidation and ignores it since there's a request being processed
(3b) Bob's cache finishes handling Alice's change, and believes it is up to date

However, it requires very specific timing between steps 3 and 4, so it doesn't seem much worth worrying about to me.

>  If we've received a refresh notification, seems like we shouldn't knowingly serve stale data.

As soon as we receive such a notification, we'll both invalidate the cache and send off a network request to get new data. However, if that request fails (with the full 15s timeout), I don't think we should make people wait the full 15 seconds the next time they want to look at their TLF list if they went offline right after receiving the notification.

Reply:
Ah I forgot about `refreshWaiting`, thanks.  I'm not sure I agree about that race you mention not being a big deal, because when it does inevitably happen it will be super confusing.  But I'm willing to table that for now.

> As soon as we receive such a notification, we'll both invalidate the cache and send off a network request to get new data

I don't see where the immediate invalidation happens.  It looks to me like it blocks on `FavoriteList` before sending anything to the main goroutine, right?  Seems like we'll keep knowingly serving stale data until `FavoriteList` returns.

> I don't think we should make people wait the full 15 seconds the next time they want to look at their TLF list if they went offline right after receiving the notification.

You could un-invalidate the cache in the case where the request failed.  That should be pretty easy.

Reply:
I see you requested another review after renaming the mode stuff, but I've been waiting for a response about my proposal about just doing the fetch in the main goroutine (blocking new favorites requests while we do the fetch), and if it fails just continue to use the previously-cached value for new requests after that, so there's only the one 15-sec blocking period.  Seems like it would be so rare that you'd go offline immediately after getting a invalidation notification that this would be the best plan.

Reply:
I've still got it on my list as TODO respond to this comment and fix stuff - not sure why I requested said review - maybe I requested it before I saw this comment. I've been acting under the assumption the ball is in my court.

Reply:
Coming back to this, I agree that it's better to wait the 15 seconds, and the chances that we're offline are very slim. I changed one call to be `FavoritesRefreshModeInMainFavoritesLoop` - I think that addresses this concern entirely. Also fixed mocks and unswallowed an error. I think this is now actually re-ready for review.